File: cpp_solutions/DynamicProgramming/Unique_Binary_Search_Trees.cpp
/*

Given n, how many structurally unique BST's (binary search trees) that store values 1...n?

For example,
Given n = 3, there are a total of 5 unique BST's.

   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3

*/


// for a certain node, the combinations equals the nums of left subtree
// multiply the nums of right subtrees.	
//
class Solution {
public:
    int numTrees(int n) {
    	vector<int> dp(n+2, 0);

    	dp[0] = 1;
    	dp[1] = 1;

    	for(int i=2; i <= n; i++){
    		for(int j=0; j<i; j++){
    			dp[i]+=dp[j]*dp[i-j-1];
    		}
    	}

    	return dp[n];
    }
};

File: python_solutions/DynamicProgramming/Unique_Binary_Search_Trees.py
class Solution:
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = [1,1,2]
        if n <= 2:
            return dp[n]
        else:
            dp += [0 for i in range(n-2)]
            for i in range(3, n+1):
                for j in range(1, i+1):
                    dp [i] += dp[i-j]*dp[j-1]
            return dp[n]